### PR TITLE
Making synthesis objects more consistent

### PR DESCRIPTION
**Pull request by billbrod**
This is a (really large) set of changes to address #135: making the synthesis objects more consistent with each other. That involves:

- making sure their call signatures and method names are as similar to each other as possible.
- making sure they all have save, load, to
- adds several validation functions they all use to make sure inputs, models, metrics look good
- a new subclass of synthesis, `OptimizedSynthesis`, which is inherited by `Geodesic`, `MADCompetition` and `Metamer`, since they all use optimizers and thus have more similarities
- metamer's coarse-to-fine synthesis is now moved to a separate subclass, `MetamerCTF`, which simplifies the internals of metamer a lot (there were `if` statements everywhere)
- all plotting functions moved out of the class so they reside in the file of the same name (e.g., `po.synth.eigendistortion.display_eigendistortion`), which follows what happened in #136 for metamer
- adds two new documentation pages, `synthesis.rst` and `models.rst`, which describe the synthesis and models, respectively. I expect models will be read by users, whereas synthesis is most helpful for us and other developers.
- adds a new notebook, `synthesis_extensions.ipynb`, which describes how to inherit synthesis objects and over-write methods for simple extensions.
- removes `Synthesis.ipynb`, as it was out of date and no longer helpful
- uses `@property` (see [here](https://realpython.com/python-property/) for an overview) to make many important properties read-only (don't want users over-writing `image` or `model` after initialization). also used to change the structure of some attributes (e.g., make sure that `.geodesic` always returns the endpoints plus the optimized part, and that `.saved_metamer` is a tensor, rather than the list we append to during synthesis), so that the user's interaction with them is always consistent, even if they hit a keyboard interrupt or hit an exception during synthesis (preventing clean-up code from being written)
- adds tests for new things
- checks ability for multi-batch / channel of all synthesis methods (and adds tests). All can accept multi-channel images, and `Metamer` and `MADCompetition` can accept multi-batch, though `MADCompetition` requires the output of the metrics to have a single value (e.g., a video metric). `Metamer` can synthesize multiple images in parallel across batch, though that's probably not a good idea
- also checks for ability to work with different dtypes (torch.float64, mostly)
- started changing generic Exceptions to more specific ones (generally, `ValueError`), though didn't finish #85 , and used more specific error catching in tests

`Geodesic` is probably changed the most by this. After making the changes I ran the new and `main` versions with the following:
```py
start = center_crop(po.load_images('data/256/einstein.pgm', False), [64])
stop = center_crop(po.load_images('data/256/curie.pgm'), [64])
model = po.simul.OnOff((31,31), pretrained=True)
model.eval()
po.tools.remove_grad(model)
po.tools.set_seed(0)
moog = po.synth.Geodesic(start, stop, model, 5)
moog.synthesize()
po.synth.geodesic.plot_loss(moog)
moog.losses[-1]
po.synth.geodesic.plot_deviation_from_line(moog)
```

(remove the `po.tools.remove_grad` line and change the plotting calls to `moog.plot_loss()`, `moog.plot_deviation_from_line()` to get this running with `main` version)

The outputs look the same, have the same loss, and the deviation plot looks the same. They're not identical (`torch.allclose` fails) and they take different number of iterations, but I believe functionality has been preserved. If we have a good working synthesis example, I'd love to add it to tests, but that might take a bit more playing around

See the new documentation page `synthesis.rst` for a description of how things are structured, that will probably help understand the changes to the synthesis methods. Also the new `models.rst` for a description of what models have to meet in order to be used.

I'm not strongly tied to any of the changes here, just the idea of consistency and I had to make several arbitrary decisions, so I'm happy to change stuff if people disagree. I'd definitely like feedback specifically on  where  the new documentation pages and notebook belong? should they be structured like this (all flat) or should we do a sort of "for developers" or "advanced usage" hierarchy for `synthesis.rst` and `synthesis_extensions.ipynb`?

### CODE REVIEW

File: plenoptic/synthesize/eigendistortion.py [hunk 1]
```diff
@@ -57,13 +60,21 @@ def fisher_info_matrix_eigenvalue(
     Fv = fisher_info_matrix_vector_product(y, x, v, dummy_vec)
 
     # compute eigenvalues for all vectors in v
-    lambduh = torch.stack([a.dot(b) for a, b in zip(v.T, Fv.T)])
-    return lambduh
+    lmbda = torch.stack([a.dot(b) for a, b in zip(v.T, Fv.T)])
```
Comment (lyndond) [inline plenoptic/synthesize/eigendistortion.py:63]: offended by this change

**Review by lyndond — APPROVED** [review]
There were too many changes for me to go through with a fine-tooth comb but overall looks reasonable. I'll go through ipynbs with the nbreviewer next

File: plenoptic/simulate/models/portilla_simoncelli.py [hunk 2]
```diff
@@ -214,10 +215,7 @@ def forward(self, image, scales=None):
             A flattened tensor (1d) containing the measured representation statistics.
 
         """
-        while image.ndimension() < 4:
```
[the remark this answers is not in the document]
Comment (billbrod) [reply plenoptic/simulate/models/portilla_simoncelli.py:217]: you're right, I should use the new validate input function I wrote.

File: plenoptic/simulate/models/naive.py [hunk 3]
```diff
@@ -5,8 +5,8 @@
 import numpy as np
 from torch.nn import functional as F
 
-from plenoptic.tools.conv import same_padding
-from plenoptic.simulate.canonical_computations.filters import circular_gaussian2d
+from ...tools.conv import same_padding
```
[the remark this answers is not in the document]
Comment (billbrod) [reply plenoptic/simulate/models/naive.py:8]: I switched to this because there can be some ambiguity with absolute imports: if you install `plenoptic` in an environment using pip, then clone the repo and try to use the local version, you'll have two versions of plenoptic, and absolute imports default to the pip-installed one (depends on your `sys.path`, I believe). I'm pretty sure this happened to me while working on something in pyrtools. This might only come up during development, but relative imports will always use the corresponding version, regardless of `sys.path` (and is more resilient to restructuring), so I think it's worth the slightly more confusing import statement

File: plenoptic/simulate/models/portilla_simoncelli.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (billbrod) [reply plenoptic/simulate/models/portilla_simoncelli.py:217]: you mean, is it worth using the whole `validate_input` function? Right now, PS doesn't support batch, and I do think the dtype checking is worthwhile as well

**Comment by billbrod** [comment]
Here's the docs build for [this PR](https://plenoptic--196.org.readthedocs.build/en/196/). The new pages are [

**Comment by billbrod** [comment]
The docs build for this PR is [here](https://plenoptic--196.org.readthedocs.build/en/196/). The new pages are [Model Requirements](https://plenoptic--196.org.readthedocs.build/en/196/models.html) (user-facing) and [Synthesis objects](https://plenoptic--196.org.readthedocs.build/en/196/synthesis.html) (developer-facing).

**Comment by pehf** [comment]
Thanks for maintaining this codebase! Here are a few comments from examining the files relevant to geodesics, there were too many changes for me to look at all of it in detail.

The ordering of methods is un-intuitive to me,  in my opinion sprinkling software technicalities/abstractions throughout risks obscuring the few lines where the action/computation is happening - which could make this file hard to read. Is there a good way to minimize this?

The `@property` decorator is new to me, and I don't really understand its purpose here - all good if it makes your life easier when maintaining this software.

The criterion of `check convergence` was confusing to me at first, but mostly I wonder why it is copied and used with a different name - why not directly call the pixel_change_convergence function inside of the synthesis method?

Why not apply the `po.tools.remove_grad` inside of the class, maybe at initialization, so that the user doesn't need to remember doing this extra step?

Maybe replace my clunky finite difference function with the built-in `torch.diff` function.

Note that some of the core functionalities concerning geodesics currently reside in the `geodesics improvements` branch (eg. conditional geodesic, and PC visualizations, as well as testing on factorized steerable pyramids). It would be great to integrate these pieces in the revamped library.

**Comment by billbrod** [comment]
Thanks for the feedback @pehf. For each point:

- Agreed. I rearranged them and tried to write up my thoughts in `synthesis.rst`
- `@property` does two things, which I think are worth having:
   - makes the important public attributes (e.g., `image`, the synthesis attribute) read-only, so the user can't accidentally over-write them.
   - allows us to have different internal and external representations. E.g., `_geodesic` just includes the frames of the geodesic that we optimize, while `geodesic` includes the endpoints. Previously, this was handled by calling some clean up code at the end of `synthesize`, but that was fragile: if an exception was hit or the user stopped in the middle of execution, that code wouldn't run and it would be confusing.
- Agreed it's a little unclear, but I think it's worth it because it gives each of the optimized synthesis a single method the user can overwrite if they want to change the convergence criterion. Just calling e.g., `pixel_change_convergence` in the middle of the synthesis loop would require users to overwrite the whole `synthesize` method to change it. 
- I'm trying to move away from silently doing things for the user. If it looks like one of our assumptions was violated, I want to raise an exception and tell the user how to fix it, rather than try to handle it for them. Same logic for why we raise an exception is not 4d, rather than just unsqueezing it. It will help users develop a better mental model of how the code works as well as make maintenance easier -- I don't want half the users realizing models can't have gradients attached and half not realizing that.
- Yup, replaced `geodesic._finite_difference` with `torch.diff`
- And agreed, I added some notes in the notebook and docstring that geodesic is under construction, and we'll try and get those merged in in June.
